rfcs/0012-gossip.md

```diff
@@ -0,0 +1,48 @@
+# Gossip, A.K.A. Cluster Info
+
+This RFC describes the Solana's gossip protocol, also known as the network control plane.
+
+Nodes continuously share signed data objects among themselves in order to find each other, structure the network data plane, prevent censoring, etc.
+
+The basic mechanism is a 10Hz loop wherein every node sends a push message and/or a pull message.  Push and pull messages may elicit responses, and push messages may be forwarded on to others in the network.
+
+Gossip runs on a well-known UDP/IP port, or a port in a well-known range.  Once a network is bootstrapped, nodes advertise to each other where to find their gossip endpoint (a socket address).
+
+## Gossip Records
+
+Records shared over gossip are arbitrary, but signed and versioned (with a timestamp), and need to make sense to the node receiving them.  A strategy of "most recently recevied wins" is employed to manage what's stored in cluster info.
+
+## Push
+
+### Push Message
+
+A push message is "hey network, I have something new to share".  Nodes send push messages to a `fanout` number of push peers.
+
+Upon receiving a push message, a node examines the message for:
+
+1. duplication: if the message has been seen before, the node responds with a "prune" message drops the message.
+
+2. new info: if the message is new the node
+   a. stores the new information with an updated version in its cluster info
+   b. stores the message in "pushed once" (used for detecting duplicates)
+   c. retransmits the messages to its own push peers
+
+3. timeout: a push message has a time to live, beyond which the message is dropped
+
+### Push Peers, Prune Message
+
+A nodes selects its push peers at random from the set of known peers.  The node keeps this selection for a relatively long time.  When a prune message is received, the node drops the push peer that sent the prune.  Prune is an indication that there is another, faster path to that node than direct push.
+
+A new set of push peers is selected (TODO: when is that??).
```

Comment:
rotate 1 node every `CRDS_GOSSIP_PULL_CRDS_TIMEOUT_MS/2`

Reply:
sounds good.  looks like this code isn't present today...

Reply:
@rob-solana https://github.com/solana-labs/solana/blob/master/src/cluster_info.rs#L96

Reply:
yeah, but that's apparently only called once per cluster_info instantiation.  also, it's a complete reboot of push peers, right?

Reply:
That's not a complete reboot of push peers?

Reply:
https://github.com/solana-labs/solana/blob/master/src/cluster_info.rs#L571

Reply:
@rob-solana its not, it just reboots 1.  the number is a denominator of how many to refresh.